libs/context/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+use std::{
+    any::{Any, TypeId},
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+#[derive(Debug, Default)]
+pub struct Context {
+    store: HashMap<(String, TypeId), Box<dyn Any>>,
```

Comment:
Could we adopt something different than a `String` as a key type?
I'd strive to avoid too many memory allocations just to retrieve dynamically populated values from `Context`. Could the key be represented a closed set type, like a numeric `enum`, whose static list is append-only (similar to the introspection warning codes list we had some months ago)?

E.g.:

```rust
enum ContextKey {
  TRACING,
  PRISMA_QUERY,
  ...
}
```

and

```rust
#[derive(Debug, Default)]
pub struct Context {
    store: HashMap<(ContextKey, TypeId), Box<dyn Any>>,
}
```

Reply:
One can also try to extend this idea to `TypeId` as well, but I think that can happen in a future query and only if we see a noticeable slowdown in benchmarked performance.

Reply:
No we can't sorry, if we had an enum then the enum has to be known by the context, thus making it coupled to it. On top of that if this is passed down as owned value it won't be cloned so there would only be one allocation per key (in case it's not wrapped in an Arc<Mutex>)

What I will do in case needed, is to make it receive a &'static str as a key, as the users of context will identify the keys by name, but rather not prematurely optimize anything until I see how this is applied when refactoring (I have one in progress)

Reply:
I see. I was recommending intentionally coupling Context with its available ContextKeys for the sake of avoiding String allocations, which have already bit us several times in the past: we risk using `String`s now just to replace them later. Considering this is a library for the Query Engine and not a general purpose crate, I’d find the coupling acceptable